crates/spawner/src/destroyer.rs

```diff
@@ -1,33 +1,253 @@
+use std::marker::PhantomData;
+
+use bevy::ecs::query::{ReadOnlyWorldQuery, WorldQuery};
 use bevy::prelude::*;
 use de_core::{baseset::GameSet, objects::ObjectType, player::Player, state::AppState};
 use de_objects::Health;
 
-use crate::{ObjectCounter, SpawnerSet};
+use crate::ObjectCounter;
 
-pub(crate) struct DestroyerPlugin;
+pub(crate) struct DespawnerPlugin;
 
-impl Plugin for DestroyerPlugin {
+impl Plugin for DespawnerPlugin {
     fn build(&self, app: &mut App) {
-        app.add_system(
-            destroy
-                .in_base_set(GameSet::Update)
-                .run_if(in_state(AppState::InGame))
-                .in_set(SpawnerSet::Destroyer),
-        );
+        app.add_event::<DespawnEvent>()
+            .add_system(
+                find_dead
+                    .in_base_set(GameSet::Update)
+                    .run_if(in_state(AppState::InGame))
+                    .before(DespawnerSet::Despawn),
+            )
+            .add_system(
+                despawn
+                    .in_base_set(GameSet::Update)
+                    .run_if(in_state(AppState::InGame))
+                    .in_set(DespawnerSet::Despawn),
+            );
     }
 }
 
-fn destroy(
-    mut commands: Commands,
+#[derive(Debug, Clone, PartialEq, Eq, Hash, SystemSet)]
+pub enum DespawnerSet {
+    /// This set is run before the despawning systems run (If you expect units to die, you should
+    /// run your system in this set and the `Update` base set)
+    Destruction,
+    Despawn,
+}
+
+struct DespawnEvent(Entity);
+
+// Find all entities with low health and mark them for despawning
+fn find_dead(
     mut counter: ResMut<ObjectCounter>,
     entities: Query<(Entity, &Player, &ObjectType, &Health), Changed<Health>>,
+    mut event_writer: EventWriter<DespawnEvent>,
 ) {
     for (entity, &player, &object_type, health) in entities.iter() {
         if health.destroyed() {
             if let ObjectType::Active(active_type) = object_type {
                 counter.player_mut(player).unwrap().update(active_type, -1);
             }
-            commands.entity(entity).despawn_recursive();
+            event_writer.send(DespawnEvent(entity));
+        }
+    }
+}
+
+fn despawn(mut commands: Commands, mut despawning: EventReader<DespawnEvent>) {
+    for entity in despawning.iter() {
+        commands.entity(entity.0).despawn_recursive();
+    }
+}
+
+#[derive(Debug)]
+pub struct DespawnEventsPlugin<Q, T, F = ()>
```

Comment:
I would add docs, something like: ``/// Add this plugin to receive [`DespawnedComponentsEvent`] when matching entities are despawned.``